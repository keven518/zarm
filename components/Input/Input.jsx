import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import InputNumber from './InputNumber';
import InputText from './InputText';
import InputTextarea from './InputTextarea';

class Input extends PureComponent {

  render() {
    const { type, autosize, ...others } = this.props;

    switch (type) {
      case 'number':
        return <InputNumber ref={(ele) => { this.input = ele; }} {...others} />;

      case 'textarea':
        return <InputTextarea ref={(ele) => { this.input = ele; }} autosize={autosize} {...others} />;

      default:
        return <InputText ref={(ele) => { this.input = ele; }} {...others} />;
    }
  }
}

Input.propTypes = {
  type: PropTypes.string,
};

Input.defaultProps = {
  type: 'text',
};

export default Input;
