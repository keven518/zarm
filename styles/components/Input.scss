$prefixCls: $prefixCls-input;

.#{$prefixCls} {
  position: relative;
  width: 100%;
  padding: r(12) 0;

  input,
  textarea {
    position: relative;
    display: inline-block;
    width: 100%;
    padding: 0;
    background-color: #fff;
    background-image: none;
    border-width: 0;
    outline: 0;
    -webkit-appearance: none;

    &[type="date"],
    &[type="text"] {
      height: r(21);
    }

    &::-webkit-input-placeholder {
      color: $placeholder-color;
    }

    &:focus {
      border-color: #57c5f7;
      outline: 0;
    }

    &:hover {
      border-color: #57c5f7;
    }

    &.disabled,
    &:disabled {
      cursor: not-allowed;
      opacity: $disabled-opacity;
      /* background-color: #f3f5f7; */

      &:focus,
      &.focus,
      &:hover {
        border-color: #ccc;
      }
    }
  }

  textarea {
    line-height: 1.42857143;
    overflow: auto;
    resize: none;
    word-break: break-all;
    word-wrap: break-word;
  }

  &-placeholder {
    position: absolute;
    left: 0;
    top: 0;
    color: #c1c1c1;
  }

  &-length {
    text-align: right;
    color: #c1c1c1;
    font-size: r(12);
  }

  &-number-picker {
    background-color: #fff;
  }
}
