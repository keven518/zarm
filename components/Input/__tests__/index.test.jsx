import React from 'react';
import { render, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Input from '../index';

describe('Input', () => {
  it('renders correctly', () => {
    const wrapper = render(<Input type="text" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('type is textarea renders correctly', () => {
    const wrapper = render(<Input type="textarea" rows={4} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('type is date renders correctly', () => {
    const wrapper = render(<Input type="date" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('autosize', () => {
    jest.useFakeTimers();
    const props = {
      autosize: true,
      type: 'textarea',
      rows: 4,
      value: 'foo',
      onChange: jest.fn(),
    };
    const wrapper = mount(<Input {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.find('textarea').simulate('change', { target: { value: 'this is a test!' } });
    expect(props.onChange).toBeCalledWith(expect.objectContaining({ target: { value: 'this is a test!' } }));
    jest.runAllTimers();
    wrapper.unmount();
  });

  it('type text has autosize', () => {
    jest.useFakeTimers();
    const wrapper = mount(<Input autosize type="text" />);
    wrapper.find('input').simulate('change', { target: { value: 'this is a test!' } });
    jest.runAllTimers();
  });

  it('showLength', () => {
    const wrapper = render(<Input showLength maxLength={100} type="textarea" rows={4} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
