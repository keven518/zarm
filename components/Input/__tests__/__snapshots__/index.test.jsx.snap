// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input autosize 1`] = `
<Input
  autosize={true}
  onChange={[Function]}
  rows={4}
  type="textarea"
  value="foo"
>
  <InputTextarea
    autosize={true}
    disabled={false}
    onChange={[Function]}
    prefixCls="za-input"
    rows={4}
    showLength={false}
    value="foo"
  >
    <div
      className="za-input za-input-textarea"
    >
      <textarea
        disabled={false}
        onChange={[Function]}
        rows={4}
        value="foo"
      />
    </div>
  </InputTextarea>
</Input>
`;

exports[`Input renders correctly 1`] = `
<div
  class="za-input za-input-text"
>
  <input
    type="text"
  />
</div>
`;

exports[`Input showLength 1`] = `
<div
  class="za-input za-input-textarea"
>
  <textarea
    maxlength="100"
    rows="4"
  />
  <div
    class="za-input-length"
  >
    0/100
  </div>
</div>
`;

exports[`Input type is date renders correctly 1`] = `
<div
  class="za-input za-input-text"
>
  <input
    type="text"
  />
</div>
`;

exports[`Input type is textarea renders correctly 1`] = `
<div
  class="za-input za-input-textarea"
>
  <textarea
    rows="4"
  />
</div>
`;
